[Cocoa] Setting to drop all assertions immediately

salinas-miguel opened the pull request:
#### a5486e6679b8f6ca940f78aa8ec427f5607fc64c
<pre>
[Cocoa] Setting to drop all assertions immediately
<a href="https://bugs.webkit.org/show_bug.cgi?id=253910">https://bugs.webkit.org/show_bug.cgi?id=253910</a>
rdar://problem/106722788

Reviewed by Chris Dumez.

Having an option to quickly remove all assertions would really help
with testing background process suspension

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultShouldDropSuspendedAssertionAfterDelay):
(WebKit::defaultShouldTakeSuspendedAssertions): Deleted.
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::ProcessThrottler):
(WebKit::ProcessThrottler::setThrottleState):
(WebKit::ProcessThrottler::dropSuspendedAssertionTimerFired):
(WebKit::ProcessThrottler::delaySuspension):
(WebKit::ProcessThrottler::removeAllAssertionsTimerFired): Deleted.
* Source/WebKit/UIProcess/ProcessThrottler.h:
(WebKit::ProcessThrottler::setShouldDropSuspendedAssertionAfterDelay):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldDropSuspendedAssertionAfterDelay const):
(WebKit::WebProcessProxy::addExistingWebPage):
(WebKit::WebProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/WebProcessProxy.h:

Canonical link: <a href="https://commits.webkit.org/261886@main">https://commits.webkit.org/261886@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/133efe0d8685d3886280c62139c6b0cb26c5fd78

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112750 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21938 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1453 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4559 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121294 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23281 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13104 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5726 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118519 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17297 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100550 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105865 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99256 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1077 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46315 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/101056 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14245 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1118 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/12379 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14943 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10466 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/102573 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20262 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53110 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/32018 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8302 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16790 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/110616 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/27320 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -191,8 +191,10 @@ void ProcessThrottler::setThrottleState(ProcessThrottleState newState)
     });
 
-    if (m_assertion->type() == ProcessAssertionType::Suspended)
+    if (m_assertion->type() == ProcessAssertionType::Suspended && !m_shouldDropSuspendedAssertionImmediately)
```
I think I would have factored this like so:
```
if (m_assertion->type() == ProcessAssertionType::Suspended) {
    if (m_shouldDropSuspendedAssertionImmediately)
        m_assertion = nullptr;
    else
        m_removeAllAssertionsTimer.startOneShot(removeAllAssertionsTimeout);
} else
    m_removeAllAssertionsTimer.stop();
```

To avoid checking `m_assertion->type() == ProcessAssertionType::Suspended` twice.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -691,6 +691,15 @@ bool WebProcessProxy::shouldTakeSuspendedAssertion() const
 
+bool WebProcessProxy::shouldDropSuspendedAssertionImmediately() const
+{
+    for (auto& page : m_pageMap.values()) {
```
WTF::anyOf() is your friend here:
```
return WTF::anyOf(m_pageMap.values(), [](auto& page) {
    return page->preferences().shouldDropSuspendedAssertionImmediately();
});
```

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1207,6 +1217,7 @@ void WebProcessProxy::didFinishLaunching(ProcessLauncher* launcher, IPC::Connect
 #endif // USE(RUNNINGBOARD)
 
     m_throttler.setShouldTakeSuspendedAssertion(shouldTakeSuspendedAssertion());
```
I think these two settings are too confusing with this proposal. It seems like not taking suspended assertions should be identical to dropping suspended assertions immediately.

My proposal:
- shouldTakeSuspendedAssertion setting:
  * ON: Takes a Suspended assertion when the process throttler goes to suspended state (current behavior)
  * OFF: Do not take any assertions when the process throttler goes to suspended state (new behavior, basically what you're doing in this patch behind another setting.

- shouldDropSuspendedAssertionAfterDelay:
  * ON: Takes a Suspended assertion when the process throttler goes to suspended state and drops the assertion when the timer fires (new behavior, This used to be the shouldTakeSuspendedAssertion=OFF behavior)
  * OFF: Don't start a timer to drop the Suspended assertion when the process throttler goes to suspended state

cdumez reviewed (approved):
LGTM, much nicer IMO.

salinas-miguel committed:
[Cocoa] Setting to drop all assertions immediately
https://bugs.webkit.org/show_bug.cgi?id=253910
rdar://problem/106722788

Reviewed by Chris Dumez.

Having an option to quickly remove all assertions would really help
with testing background process suspension

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultShouldDropSuspendedAssertionAfterDelay):
(WebKit::defaultShouldTakeSuspendedAssertions): Deleted.
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::ProcessThrottler):
(WebKit::ProcessThrottler::setThrottleState):
(WebKit::ProcessThrottler::dropSuspendedAssertionTimerFired):
(WebKit::ProcessThrottler::delaySuspension):
(WebKit::ProcessThrottler::removeAllAssertionsTimerFired): Deleted.
* Source/WebKit/UIProcess/ProcessThrottler.h:
(WebKit::ProcessThrottler::setShouldDropSuspendedAssertionAfterDelay):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldDropSuspendedAssertionAfterDelay const):
(WebKit::WebProcessProxy::addExistingWebPage):
(WebKit::WebProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/WebProcessProxy.h:

Canonical link: https://commits.webkit.org/261886@main
